lib/webpage_cache_service/cache.rb

```diff
@@ -79,42 +83,67 @@ def client
     def get uri:
       response = client.get uri
 
-      content_type = response.content_type.mime_type
+      # Without manually managing the temp file, there isn't an easy way to
+      # breakup the following logic but I did my best and fuck you too rubocop
+      attachments = temp_file_for uri: uri do |temp_file, uri_filename|
+        write_body response: response, to: temp_file
 
-      temp_key = Digest::SHA256.hexdigest uri.to_s
-      attachment = Tempfile.create temp_key do |temp_file|
-        temp_file.binmode
+        offline_cache.assets.attach(
+          io: temp_file,
+          filename: uri_filename,
+          content_type: get_content_type(response: response, io: temp_file),
+          metadata: build_metadata(response: response)
+        )
+      end
+
+      [ response, attachments.first ]
+    end
 
-        response.body.each do |partial|
-          temp_file.write partial
-        end
+    def temp_file_for uri:
+      uri_filename = Digest::SHA256.hexdigest uri.to_s
 
-        temp_file.rewind
+      Tempfile.create uri_filename do |temp_file|
+        temp_file.binmode
 
-        unless content_type
-          # some websites *cough*offline.pink*cough* don't return a content type
-          # header, so we'll first see if the html doctype string is present and
-          # guess this is text/html or we'll fallback to assuming its binary
-          # https://en.wikipedia.org/wiki/Content_sniffing
-          content_type ||= MimeMagic.by_magic(temp_file)
-          content_type ||= "application/octet-stream"
+        yield temp_file, uri_filename
+      end
+    end
 
-          temp_file.rewind
-        end
+    # Streams the response body into the temp file to hopefully avoid some
+    # memory issues if this is a large document
+    def write_body response:, to:
+      io_handle = to
 
-        offline_cache.assets.attach(
-          io: temp_file,
-          filename: Digest::SHA256.hexdigest(uri.to_s),
-          content_type: content_type,
-          metadata: {
-            uri: uri.to_s,
-            status_code: response.code,
-            headers: response.headers.to_h
-          }
-        )&.first
+      response.body.each do |partial|
+        io_handle.write partial
       end
 
-      [ response, attachment ]
+      io_handle.rewind
```

Comment:
hm?

Reply:
Cause passing the io object into activestorage needs it to be rewound otherwise it writes an empty file and you end up with weird things like ```  Disk Storage (0.1ms) Downloaded file from key: 4LukgP49udYsbuMarJej4eYi
  Disk Storage (0.2ms) Deleted file from key: SwXXUd7bpv5mjgoG29HyoqMF
  Disk Storage (16.3ms) Uploaded file to key: SwXXUd7bpv5mjgoG29HyoqMF (checksum: 1B2M2Y8AsgTpgAmY7PhCfg==)
ActiveStorage::IntegrityError: ActiveStorage::IntegrityError```